tuf/builder.go

```diff
@@ -0,0 +1,402 @@
+		return rb.LoadDelegation(roleName, content, minVersion)
+	}
+}
+
+// LoadRoot loads a root if one has not been loaded
+func (rb *repoBuilder) LoadRoot(content []byte, minVersion int) error {
+	roleName := data.CanonicalRootRole
+	switch {
+	case rb.repo.Root != nil:
+		return ErrInvalidBuilderInput{"msg: root has already been loaded"}
+	case rb.failed:
+		return ErrBuildFailed{}
+	case rb.finished:
+		return ErrBuildDone{}
+	}
+
+	signedObj, err := rb.bytesToSigned(roleName, content, rb.rootChecksummer)
+	if err != nil {
+		return err
+	}
+
+	if err := rb.verifyPinnedTrust(signedObj); err != nil {
+		return err
+	}
+
+	// verify that the metadata structure is correct - we need this in order to get
+	// the root role to verify that signatures are self-consistent
+	signedRoot, err := data.RootFromSigned(signedObj)
+	if err != nil {
+		return err
+	}
+
+	rootRole, err := signedRoot.BuildBaseRole(roleName)
+	if err != nil { // this should never happen, since it's already been validated
+		return err
+	}
+	// validate that the signatures for the root are consistent with its own definitions
+	if err := signed.Verify(signedObj, rootRole, minVersion); err != nil {
+		return err
+	}
+
+	rb.repo.SetRoot(signedRoot)
+	return nil
+}
+
+// validate against old keys or pinned trust certs
+func (rb *repoBuilder) verifyPinnedTrust(signedObj *data.Signed) error {
+	if rb.rootRole == nil {
+		// TODO: certs.ValidateRoot should only check the trust pinning - we will
+		// validate that the root is self-consistent with itself later
+		// it also calls RootToSigned, so there are some inefficiencies here
+		if err := certs.ValidateRoot(rb.certStore, signedObj, rb.gun); err != nil {
+			logrus.Debug("TUF repo builder: root failed validation against trust certificates")
+			return err
+		}
+	} else {
+		// verify with existing keys rather than trust pinning
+		if err := signed.VerifySignatures(signedObj, *rb.rootRole); err != nil {
+			logrus.Debug("TUF repo builder: root failed validation against previous root keys")
+			return err
+		}
+	}
+	return nil
+}
+
+func (rb *repoBuilder) LoadTimestamp(content []byte, minVersion int) error {
+	roleName := data.CanonicalTimestampRole
+	switch {
+	case rb.repo.Timestamp != nil:
+		return ErrInvalidBuilderInput{msg: "timestamp has already been loaded"}
+	case rb.repo.Root == nil:
+		return ErrInvalidBuilderInput{msg: "root must be loaded first"}
+	case rb.failed:
+		return ErrBuildFailed{}
+	case rb.finished:
+		return ErrBuildDone{}
+	}
+	timestampRole, err := rb.repo.Root.BuildBaseRole(roleName)
+	if err != nil { // this should never happen, since it's already been validated
+		return err
+	}
+
+	signedObj, err := rb.bytesToSignedAndValidateSigs(timestampRole, content, minVersion)
+	if err != nil {
+		return err
+	}
+
+	signedTimestamp, err := data.TimestampFromSigned(signedObj)
+	if err != nil {
+		return err
+	}
+
+	rb.repo.SetTimestamp(signedTimestamp)
+	return rb.validateChecksums(roleName, signedTimestamp.Signed.Meta)
```

Comment:
I originally had it that way, but for snapshot, we need to set the failed snapshot because otherwise we won't be able to bootstrap a new builder (since the snapshot wouldn't ever be set).  Because if the checksums are wrong, we need to download a new root based on the snapshot that we downloaded.  :|

Another way to do this would be to have the `tuf/client` ALSO keep all the metadata that it's downloaded, so it has a cache of metadata separate from the builder so it can bootstrap a new builder (although it would also need to store the certificate validator, etc. that is right now created in `client/client` in order to make a completely new builder).